TESTS/usb_device/serial/main.cpp

```diff
@@ -0,0 +1,538 @@
+ */
+void test_serial_getc()
+{
+    USBSerial usb_serial(false, USB_SERIAL_VID, USB_SERIAL_PID);
+    usb_serial.connect();
+    if (send_testcase_data(MSG_KEY_SEND_BYTES_SINGLE, USB_SERIAL_VID, USB_SERIAL_PID) == false) {
+        TEST_ASSERT_MESSAGE(0, "Dev-host communication error.");
+        return;
+    }
+    while (!usb_serial.connected()) {
+        wait_ms(1);
+    }
+    for (int expected = 0xff; expected >= 0; expected--) {
+        TEST_ASSERT_EQUAL_INT(expected, usb_serial.getc());
+    }
+    for (int expected = 0; expected <= 0xff; expected++) {
+        TEST_ASSERT_EQUAL_INT(expected, usb_serial.getc());
+    }
+    // Wait for the host to close its port.
+    while (usb_serial.ready()) {
+        wait_ms(1);
+    }
+    usb_serial.disconnect();
+}
+
+/** Test Serial printf & scanf
+ *
+ * Given the USB Serial device connected to a host
+ * When the device trensmits a formatted string with a random value
+ *     using the printf method
+ *     and the host sends it back to the device
+ * Then the device can successfully read the value using scanf method
+ *     and the value received is equal value sent
+ */
+void test_serial_printf_scanf()
+{
+    USBSerial usb_serial(false, USB_SERIAL_VID, USB_SERIAL_PID);
+    usb_serial.connect();
+    if (!send_testcase_data(MSG_KEY_LOOPBACK, USB_SERIAL_VID, USB_SERIAL_PID)) {
+        TEST_ASSERT_MESSAGE(0, "Dev-host communication error.");
+        return;
+    }
+    while (!usb_serial.connected()) {
+        wait_ms(1);
+    }
+    static const char fmt[] = "Formatted\nstring %i.";
+    int tx_val, rx_val, rc;
+    for (int i = 0; i < SERIAL_LOOPBACK_REPS; i++) {
+        tx_val = rand();
+        rx_val = tx_val + 1;
+        rc = usb_serial.printf(fmt, tx_val);
+        TEST_ASSERT(rc > 0);
+        rc = usb_serial.scanf(fmt, &rx_val);
+        TEST_ASSERT(rc == 1);
+        TEST_ASSERT_EQUAL_INT(tx_val, rx_val);
+    }
+    // Wait for the host to close its port.
+    while (usb_serial.ready()) {
+        wait_ms(1);
+    }
+    usb_serial.disconnect();
+}
+
+utest::v1::status_t testsuite_setup(const size_t number_of_cases)
+{
+    GREENTEA_SETUP(25, "usb_device_serial");
+    srand((unsigned) ticker_read_us(get_us_ticker_data()));
+    return utest::v1::greentea_test_setup_handler(number_of_cases);
+}
+
+Case cases[] = {
+    Case("CDC USB reconnect", test_cdc_usb_reconnect),
+    Case("CDC RX single bytes", test_cdc_rx_single_bytes),
+    Case("CDC RX single bytes concurrent", test_cdc_rx_single_bytes_concurrent),
+    Case("CDC RX multiple bytes", test_cdc_rx_multiple_bytes),
+    Case("CDC RX multiple bytes concurrent", test_cdc_rx_multiple_bytes_concurrent),
+    Case("CDC loopback", test_cdc_loopback),
+    Case("Serial USB reconnect", test_serial_usb_reconnect),
+    Case("Serial terminal reopen", test_serial_term_reopen),
+    Case("Serial getc", test_serial_getc),
+    Case("Serial printf/scanf", test_serial_printf_scanf),
```

Comment:
It may also be good to test `line_coding_changed` if you are testing USBCDC in this PR.

Reply:
New test case added -- `Serial terminal reopen`.